File: leetcode/python/1405.longest-happy-string.py
class Solution:
    def longestDiverseString(self, a: int, b: int, c: int, A = 'a', B = 'b', C = 'c') -> str:
        if a < b:
            return self.longestDiverseString(b, a, c, B, A, C)
        if b < c:
            return self.longestDiverseString(a, c, b, A, C, B)
        if b == 0:
            return A * min(a, 2)
        
        aCount = min(a, 2)
        bCount = 1 if (a - aCount >= b) else 0
        return A * aCount + B * bCount + self.longestDiverseString(a - aCount, b - bCount, c, A, B, C)

File: leetcode/cpp/1405.longest-happy-string.cpp
class Solution {
public:
    string longestDiverseString(int a, int b, int c, char A='a', char B='b', char C='c') {
        if (a < b) {
            return longestDiverseString(b, a, c, B, A, C);
        }
        if (b < c) {
            return longestDiverseString(a, c, b, A, C, B);
        }
        if (b == 0) {
            return string(min(a, 2), A);
        }
        int aCount = min(a, 2);
        int bCount = (a - aCount >= b) ? 1 : 0;
        return string(aCount, A) + string(bCount, B) + longestDiverseString(a-aCount, b-bCount, c, A, B, C);
    }
};

